### ISSUE TITLE
[bug] changes to create policy not obeyed despite strict mode

### ISSUE THREAD

**Issue by borkd** [moosefs/moosefs#146]
How to replicate my environment:

- moosefs/3.0.100
chunkservers: 1 chunkserver with label "V", 2+ chunkservers with label "PSZ"
- create storage class livetest: `mfsscadmin /mnt/mfs create -a 0 -C V -K PSZ -A PSZ -d 1 livetest`
- `mkdir /mnt/mfs/livetest && mfssetsclass -rH livetest /mnt/mfs/livetest`
- start a background process writing files under /mnt/mfs/livetest/ and observe chunk numbers in servers tab and changes in resources tab.

Expectations are met: chunks are written to the "V" chunkserver and then replicated to chunkservers labeled "PSZ". During OVERLOADED state chunks are written opportunistically. 

Now change the create policy: 
`mfsscadmin /mnt/mfs modify -m strict -a 0 -C PSZ -K PSZ -A PSZ -d 1 livetest`

What is expected: number of chunks on the "V" chunkserver goes down to 0 as replicas are made; new chunks are created only on servers with labels "PSZ".

What happens: number of chunks on "V" keeps increasing. Eventually it will drop to 0, after all replication activity stops.

@xandrus - can you replicate this behavior?

Edit: mfssetsclass was erroneously lacking class name in the original issue report, fixed now.

**Comment by xandrus** [comment]
Hi,
I would like to add that we are not able to reproduce such behaviour.
Meaby there is another storage class which is also writing to V because of labels expressions are not accomplished?

![image](https://user-images.githubusercontent.com/13887484/47345517-3a42ad00-d6ab-11e8-9ffd-3caf0e7a6a4c.png)

**Comment by oxide94** [comment]
Just to mention - these chunks on second CS labeles as `V` were in trash before the test, so they can be omitted. Here is the situation after purging the trash:

![screen shot 2018-10-23 at 11 05 56 am](https://user-images.githubusercontent.com/4892057/47349378-d1136780-d6b3-11e8-8aa0-a9ca7ae86685.png)

![screen shot 2018-10-23 at 11 06 10 am](https://user-images.githubusercontent.com/4892057/47349396-d5d81b80-d6b3-11e8-816f-6c73f1d978ee.png)

**Comment by borkd** [comment]
@xandrus:
- I should not have any storage classes that are running out of required number of servers.
- All data in my filesystem is assigned to classes in **strict** mode. Built-in classes 1,2,3, and 4 were redefined to use particular labels for Create ,Keep, and Archive modes. As in #134 
- **Chunkserver label `V` is referenced only in create mode for a number of classes**, with expectation than chunks are migrated away from `V` as they transition to Keep mode.
- I have kept system quiesced for a while and most chunks should be in archive mode now, yet the chunkserver with label V still houses between 2 to 3K chunks (I've seen the number fluctuate, CGI shows constant writes, and strace confirms that).

Questions:
1. Many of the chunks on the `V` chunkserver are from a built in goal/class of "1". How is is possible they would end up there? Chunkserver `V` has one more label `H`, but not P, S, or Z.
```
# mfsscadmin /mnt/mfs/ list -l
1 : 1 ; admin_only: NO ; create_mode: STRICT ; create_labels: [PSZ] ; keep_labels: [PSZ] ; arch_labels: [PSZ] ; arch_delay: 1d
[...]
```

2. The only expected activity at the moment is transitions between keep and archive, space rebalancing between chunkservers, chunk replication as individual chunkservers vanish, and deletion of extra chunks as servers eventually rejoin. My understanding is that space rebalancing and replication happens between chunkservers labeled for particular chunk state (keep, archive) state. On a system with adequate resources chunkservers servicing Create mode would be drained relatively quickly and data would never flow back to them.

3. How is it possible that on an otherwise idle cluster, with no missing servers and no undergoal chunks, the "create-only" chunkserver with label `V` keeps being written to at up to 6K writes/min? Any storage class referencing `V` in create mode periodically reports as OVERLOADED in CGI. Load average remains low at `0.14 0.09 0.14`.

4. What is the fastest way to lookup file path based on chunk ID without running mfsfileinfo on entire namespace *on the client side*? That would help to pinpoint what is going on, and would be tremendously useful in automated test harness.

@acid-maker - can you comment on 1 and 3? This topic deserves to be well documented.

**Comment by oxide94** [comment]
@acid-maker Kuba, could you plz take a look at this issue? @borkd will be happy to provide all the necessary details. This issue may be related with #134.

Thanks in advance,
Piotr

**Comment by borkd** [comment]
Almost there.. but.. more questions. 

At first glance it all looks fine:
 
<img width="1062" alt="mfs104wip-resources" src="https://user-images.githubusercontent.com/12971584/49307083-248e8780-f4a2-11e8-91fa-714a60b37652.png">

At this point I would expect the **V,H** chunkserver to be drained and empty. One chunk is still reported to live there. Let's look closer.

<img width="784" alt="mfs104wip-servers" src="https://user-images.githubusercontent.com/12971584/49307163-57388000-f4a2-11e8-8b1d-f3f4409a35fd.png">

Upon closer examination, there are 181 chunks on that filesystem. 

Chunkserver with **V,H** labels still has 181 chunks on its filesystem:
```
CHUNKSERVER_VH# find -name *.mfs | cut -d_ -f2 | sed -e 's/^0*//g' | xargs echo
```
> 3F4946 3DC72E B2F17 13069F 33EF9D 45D2F2 24E72D 7A179 22A68D 48A875 B6E19 6BA33 37E356 1427DC A186E 32D3E8 43133A 344334 21923B 77F0D 3DBDA6 20B12B 2E9653 F482E 380B2E 32654F 223807 217254 17D8BA 4661B B7E33 4A8515 83BAA E910C 3C5E6D 2E3344 23DC8F 42397A 6221F 12373D 144067 8E680 38534A 1204F9 11C4B4 DB2DC 4493BA 27F3B3 147E6E 26E9F3 44E74D 3D7565 5E3 13F160 328A1E 3FE87B 278891 3C39FA 3E3F4B 34635B 1761E9 358142 188DB9 3DD280 155F45 2220C0 2B9C90 11666D 4A43E 3D7A45 37C729 3BC5C8 37F1DC 28C048 348C85 21311B 47AEF4 3827B5 743CD 238E67 17299C 2FDDBC B21A6 48C080 2636A2 227473 4B2A5B 26C0F0 1760F 5E103 B2B0C 1AE65 3EECE 395863 17E90C 259066 48C8F6 437326 2D568F 374823 2937A6 25ADC4 33A4E9 3FD7BA 633FB 21A4CD 3FB226 30F274 64709 1CA991 48BC9C 14EECD FE273 4595D5 3416F9 2F8081 B843B 6B99B 3E4E44 EB829 63061 5BC1 22B593 433058 24156A 281F0F 3F1297 A2C26 396DDF 1066BD 23B08B 32ACC6 32E15F 551F9 2D66D5 22E55 41382F 2C0C61 6AF88 3021C6 309F39 D90D9 430A3 319FD5 22E42F 3B0C49 2FFBCB 3B15EE 301BC1 B8F38 157089 134331 3B6D7 133DF2 36BB07 C5B57 2D9198 668EF D5103 39CE2A F3C21 4A84A4 31C8FD E0A75 E450B 3A5313 18FFDE 47B7D2 FF87B 623B2 100731 BA019 2502DD EBC7A 138F75 13EDBC 28B87F 1069DF FA25F 48EFA2 4BE99A


Even if they are no longer needed and slated for removal, how come they are not accounted for and reported anywhere?

<img width="1187" alt="mfs104wip-chunk state matrix" src="https://user-images.githubusercontent.com/12971584/49307061-0e80c700-f4a2-11e8-8301-519e0f168526.png">

**Comment by borkd** [comment]
@acid-maker - I have tested 8ab9f05 for close to 5 days - the original issue seems to be resolved. I will open another ticket regarding chunks which quietly were left behind, and occupy disk space without being accounted for.

**Comment by borkd** [comment]
Chunk `21311B` from the above list is referenced by four filesystem entries: one for live data and by three separate snapshots. `mfsfileinfo` against each file reports VALID chunks on a number of chunkservers, none of which is chunkserver `.1.35` with labels 'H,V', where the chunk is squatting since 2018-10-18.

**Comment by borkd** [comment]
After checking that each of the 180 chunks had replicas elsewhere, I have restarted the chunkserver. No changes were observed. Only after I stopped it again, manually removed the `.chunkdb` file and started chunkserver disk scan has identified previously unaccounted for extra copies, and they were subsequently removed.

While it may be difficult to backtrack and determine the root cause of why those 180 chunks went MIA, there is a chance to make a small improvement in how the process is handled going forward.

Should `.chunkdb` have a sensible shelf life, after which chunk scan is forced, akin to how e2fsck behaves when the filesystem hasn't been checked for over 180 days? If so what would be a sensible value?

**Comment by jkiebzak** [comment]
+1Setting per config file



null

**Comment by acid-maker** [comment]
I don't like idea of simple aging of `.chunkdb`. Maybe chunkserver shouldn't "trust" in content of this file and use it just after start, but also run background scanning that should check if everything is fine.
Also it might be good idea to add regular background scanning to chunkservers - like once a week (this might be defined in config file).

**Comment by jkiebzak** [comment]
Any of these options, would be good. I would be in favor of the least
amount of code. One benefit of trusting this file during startup is speed,
especially with larger disks (avoiding the need to crawl 8, 10, 12 TB disks
during daemon start).

**Comment by borkd** [comment]
I'm all in for a smaller code base and empowering admins to make the final call.

**Comment by borkd** [comment]
Would it also make sense to embargo chunk deletion until chunk scan completes for any disk where hdd_folder_fastscan returned 0?

**Comment by borkd** [comment]
OK, https://github.com/moosefs/moosefs/commit/8ab9f05098f8abcda75ea83f9552154b50fe3b6c fixes the issue when testing with the new filesystem. Should we break off the feature to perform lazy, but full chunk scan after restart with .chunkdb into separate ticket, or leave this one open?

**Comment by borkd** [comment]
Leaving open until 3.0.104 is tagged, and available as a "release" on GH.

**Comment by oxide94** [comment]
MooseFS 3.0.104 [has been released](https://github.com/moosefs/moosefs/releases/tag/v3.0.104), closing.
